<a href="https://colab.research.google.com/github/syuhaza/Training-ML/blob/master/python109_nursyuhada.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Salaries Exercise

Click [here](https://drive.google.com/file/d/1CeOKg0G1_CpECEWvze30fMdbuEspkmWD/view?usp=sharing) to download the file

**Import pandas as pd**

In [ ]:
import pandas as pd

**Read Salaries.csv as a dataframe called sal**

In [ ]:
sal = pd.read_csv('/content/Salaries.csv')

**Check the head of the DataFrame**

In [ ]:
sal.head()

Id       EmployeeName                                        JobTitle  \
0   1     NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1   2       GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2   3     ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3   4  CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4   5    PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   

     BasePay  OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  \
0  167411.18         0.00  400184.25       NaN  567595.43         567595.43   
1  155966.02    245131.88  137811.38       NaN  538909.28         538909.28   
2  212739.13    106088.18   16452.60       NaN  335279.91         335279.91   
3   77916.00     56120.71  198306.90       NaN  332343.61         332343.61   
4  134401.60      9737.00  182234.59       NaN  326373.19         326373.19   

   Year  Notes         Agency  Status  
0  2011    NaN  San Francisco     NaN  
1  2011    NaN  San Francisco     NaN  
2  2011    NaN  San Francisco     NaN  
3  2011    NaN  San Francisco     NaN  
4  2011    NaN  San Francisco     NaN

**Use the .info() method to find out how many entries there are**

In [ ]:
sal.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40409 entries, 0 to 40408
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Id                40409 non-null  int64  
 1   EmployeeName      40409 non-null  object 
 2   JobTitle          40409 non-null  object 
 3   BasePay           40408 non-null  float64
 4   OvertimePay       40408 non-null  float64
 5   OtherPay          40408 non-null  float64
 6   Benefits          4249 non-null   float64
 7   TotalPay          40408 non-null  float64
 8   TotalPayBenefits  40408 non-null  float64
 9   Year              40408 non-null  float64
 10  Notes             0 non-null      float64
 11  Agency            40408 non-null  object 
 12  Status            0 non-null      float64
dtypes: float64(9), int64(1), object(3)
memory usage: 4.0+ MB


**Find the average BasePay ?**

Answer: 66325.44884050643

In [ ]:
sal['BasePay'].mean()

66325.4488404877

**Find is the highest amount of OvertimePay in the dataset ?**

Answer: 245131.88

In [ ]:
sal['OvertimePay'].max()

245131.88

**Find the job title of  JOSEPH DRISCOLL ?**

Note: Use all caps, otherwise you may get an answer that doesn't match up (there is also a lowercase Joseph Driscoll)

Answer: CAPTAIN, FIRE SUPPRESSION

In [ ]:
sal[sal['EmployeeName'] == 'JOSEPH DRISCOLL']['JobTitle']

24    CAPTAIN, FIRE SUPPRESSION
Name: JobTitle, dtype: object

**How much does JOSEPH DRISCOLL make (including benefits)?**

Answer: 270324.91

In [ ]:
sal[sal['EmployeeName'] == 'JOSEPH DRISCOLL']['TotalPayBenefits']

24    270324.91
Name: TotalPayBenefits, dtype: float64

**What is the name of highest paid person (including benefits) ?**

Note: Print the entire row of data as follows

Answer:



```
1 | NATHANIEL FORD | GENERAL MANAGER - METROPOLITAN TRANSIT...
```




In [ ]:
sal.loc[sal['TotalPay'] == sal.describe().loc['max', 'TotalPay']]

Id    EmployeeName                                        JobTitle  \
0   1  NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   

     BasePay  OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  \
0  167411.18          0.0  400184.25       NaN  567595.43         567595.43   

   Year  Notes         Agency  Status  
0  2011    NaN  San Francisco     NaN

**What is the name of lowest paid person (including benefits)? Do you notice something strange about how much he or she is paid? If yes put it as a comment inside the given code block**

Note: Print the entire row of data as follows

Answer:

```
148654 |	Joe Lopez	| Counselor, Log Cabin Ranch...
```

In [ ]:
highestindex = sal['TotalPayBenefits'].idxmin()
highestpaid_row = sal.loc[highestindex]
formatted_output = f"{highestpaid_row['Id']} | {highestpaid_row['EmployeeName']} | {highestpaid_row['JobTitle']}"
formatted_output
# Yes

'148654 | Joe Lopez | Counselor, Log Cabin Ranch'

**What was the average (mean) BasePay of all employees per year? (2011-2014) ?**

Answer:

```
Year
2011    63595.956517
2012    65436.406857
2013    69630.030216
2014    66564.421924
Name: BasePay, dtype: float64
```

In [ ]:
sal.groupby('Year')['BasePay'].mean()

Year
2011    63595.956517
2012    65436.406857
2013    69630.030216
2014    66564.421924
Name: BasePay, dtype: float64

**How many unique job titles are there ?**

Answer: 2159

In [ ]:
len(sal['JobTitle'].unique())

2159

**What are the top 5 most common jobs ?**

Answer:

```
Transit Operator                7036
Special Nurse                   4389
Registered Nurse                3736
Public Svc Aide-Public Works    2518
Police Officer 3                2421
Name: JobTitle, dtype: int64
```

In [ ]:
sal['JobTitle'].value_counts().head(5)

JobTitle
Transit Operator                7036
Special Nurse                   4389
Registered Nurse                3736
Public Svc Aide-Public Works    2518
Police Officer 3                2421
Name: count, dtype: int64

**How many Job Titles were represented by only one person in 2013? (e.g. Job Titles with only one occurence in 2013?)**

Answer: 202

In [ ]:
uniqueJob = sal[sal['Year'] == 2013]['JobTitle'].value_counts()[sal[sal['Year'] == 2013]['JobTitle'].value_counts() == 1].index.tolist()
len(uniqueJob)

202

**How many people have the word Chief in their job title ?**

Note: This is pretty tricky

Answer: 477

In [ ]:
chiefs = sal[sal['JobTitle'].str.contains('^Chief', case=False, na=False)]
len(chiefs)

299

**Is there a correlation between length of the Job Title string and Salary ?**

In [ ]:
# Yes

# Ecommerce Purchases Exercise

Click [here](https://drive.google.com/file/d/1B27xrTYJQEzTwnWNn872yOZUljJ-WlFT/view?usp=sharing) to download the file

**Import pandas**

In [ ]:
import pandas as pd

**Read in the Ecommerce Purchases csv file and set it to a DataFrame called ecom**

In [ ]:
ecom = pd.read_csv('/content/Ecommerce Purchases.csv')

**Check the head of the DataFrame.**

In [ ]:
ecom.head()

Address    Lot AM or PM  \
0  16629 Pace Camp Apt. 448\nAlexisborough, NE 77...  46 in       PM   
1  9374 Jasmine Spurs Suite 508\nSouth John, TN 8...  28 rn       PM   
2                   Unit 0065 Box 5052\nDPO AP 27450  94 vE       PM   
3              7780 Julia Fords\nNew Stacy, WA 45798  36 vm       PM   
4  23012 Munoz Drive Suite 337\nNew Cynthia, TX 5...  20 IE       AM   

                                        Browser Info  \
0  Opera/9.56.(X11; Linux x86_64; sl-SI) Presto/2...   
1  Opera/8.93.(Windows 98; Win 9x 4.90; en-US) Pr...   
2  Mozilla/5.0 (compatible; MSIE 9.0; Windows NT ...   
3  Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_0 ...   
4  Opera/9.58.(X11; Linux x86_64; it-IT) Presto/2...   

                           Company       Credit Card CC Exp Date  \
0                  Martinez-Herman  6011929061123406       02/20   
1  Fletcher, Richards and Whitaker  3337758169645356       11/18   
2       Simpson, Williams and Pham      675957666125       08/19   
3  Williams, Marshall and Buchanan  6011578504430710       02/24   
4        Brown, Watson and Andrews  6011456623207998       10/25   

   CC Security Code                  CC Provider  \
0               900                 JCB 16 digit   
1               561                   Mastercard   
2               699                 JCB 16 digit   
3               384                     Discover   
4               678  Diners Club / Carte Blanche   

                            Email                                     Job  \
0               pdunlap@yahoo.com  Scientist, product/process development   
1              anthony41@reed.com                       Drilling engineer   
2  amymiller@morales-harrison.com                Customer service manager   
3     brent16@olson-robinson.info                       Drilling engineer   
4     christopherwright@gmail.com                             Fine artist   

        IP Address Language  Purchase Price  
0  149.146.147.205       el           98.14  
1     15.160.41.51       fr           70.73  
2   132.207.160.22       de            0.95  
3     30.250.74.19       es           78.04  
4     24.140.33.94       es           77.82

**How many rows and columns are there ?**

Answer

```
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
Address             10000 non-null object
Lot                 10000 non-null object
AM or PM            10000 non-null object
Browser Info        10000 non-null object
Company             10000 non-null object
Credit Card         10000 non-null int64
CC Exp Date         10000 non-null object
CC Security Code    10000 non-null int64
CC Provider         10000 non-null object
Email               10000 non-null object
Job                 10000 non-null object
IP Address          10000 non-null object
Language            10000 non-null object
Purchase Price      10000 non-null float64
dtypes: float64(1), int64(2), object(11)
memory usage: 1.1+ MB
```

In [ ]:
ecom.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Address           10000 non-null  object 
 1   Lot               10000 non-null  object 
 2   AM or PM          10000 non-null  object 
 3   Browser Info      10000 non-null  object 
 4   Company           10000 non-null  object 
 5   Credit Card       10000 non-null  int64  
 6   CC Exp Date       10000 non-null  object 
 7   CC Security Code  10000 non-null  int64  
 8   CC Provider       10000 non-null  object 
 9   Email             10000 non-null  object 
 10  Job               10000 non-null  object 
 11  IP Address        10000 non-null  object 
 12  Language          10000 non-null  object 
 13  Purchase Price    10000 non-null  float64
dtypes: float64(1), int64(2), object(11)
memory usage: 1.1+ MB


**What is the average Purchase Price ?**

Answer: 50.34730200000025

In [ ]:
ecom['Purchase Price'].mean()

50.347302

**What were the highest purchase prices ?**

Answer: 99.989999999999995

In [ ]:
ecom['Purchase Price'].max()

99.99

**What were the lowest purchase prices ?**

Answer: 0

In [ ]:
ecom['Purchase Price'].min()

0.0

**How many people have English 'en' as their Language of choice on the website ?**

In [ ]:
english = len(ecom.loc[ecom['Language'] == 'en'].values[0])
english

14

**How many people have the job title of "Lawyer" ?**


In [ ]:
lawyer = len(ecom.loc[ecom['Job'] == 'Lawyer'].values[0])
lawyer

14

**How many people made the purchase during the AM and how many people made the purchase during PM ?**

**(Hint: Check out [value_counts()](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.value_counts.html) )**

Answer:

```
PM    5068
AM    4932
Name: AM or PM, dtype: int64
```

In [ ]:
ecom['AM or PM'].value_counts()

AM or PM
PM    5068
AM    4932
Name: count, dtype: int64

**What are the 5 most common Job Titles ?**

Answer:

```
Interior and spatial designer    31
Lawyer                           30
Social researcher                28
Purchasing manager               27
Designer, jewellery              27
Name: Job, dtype: int64
```



In [ ]:
ecom['Job'].value_counts().head(5)

Job
Interior and spatial designer    31
Lawyer                           30
Social researcher                28
Purchasing manager               27
Designer, jewellery              27
Name: count, dtype: int64

**Someone made a purchase that came from Lot: "90 WT" , what was the Purchase Price for this transaction ?**

Answer:
```
513    75.1
Name: Purchase Price, dtype: float64
```



In [ ]:
ecom.loc[ecom['Lot'] == '90 WT', 'Purchase Price']

513    75.1
Name: Purchase Price, dtype: float64

**What is the email of the person with the following Credit Card Number: 4926535242672853**

Answer:

```
1234    bondellen@williams-garza.com
Name: Email, dtype: object
```


In [ ]:
ecom.loc[ecom['Credit Card'] == 4926535242672853, 'Email']

1234    bondellen@williams-garza.com
Name: Email, dtype: object

**How many people have American Express as their Credit Card Provider *and* made a purchase above $95 ?**

Answer

```
Address             39
Lot                 39
AM or PM            39
Browser Info        39
Company             39
Credit Card         39
CC Exp Date         39
CC Security Code    39
CC Provider         39
Email               39
Job                 39
IP Address          39
Language            39
Purchase Price      39
dtype: int64
```



In [ ]:
ecom[(ecom['CC Provider'] == 'American Express') & (ecom['Purchase Price'] > 95)].count()

Address             39
Lot                 39
AM or PM            39
Browser Info        39
Company             39
Credit Card         39
CC Exp Date         39
CC Security Code    39
CC Provider         39
Email               39
Job                 39
IP Address          39
Language            39
Purchase Price      39
dtype: int64

**Hard: How many people have a credit card that expires in 2025 ?**

Answer: 1033

In [ ]:
expired_card = ecom.loc[ecom['CC Exp Date'].str.contains('25$')].value_counts()
len(expired_card)

1033

**Hard: What are the top 5 most popular email providers/hosts (e.g. gmail.com, yahoo.com, etc...)**

Answer:

```
hotmail.com     1638
yahoo.com       1616
gmail.com       1605
smith.com         42
williams.com      37
Name: Email, dtype: int64
```



In [ ]:
ecom['Email'].apply(lambda x: x.split('@')[1]).value_counts().head(5)

Email
hotmail.com     1638
yahoo.com       1616
gmail.com       1605
smith.com         42
williams.com      37
Name: count, dtype: int64